### DISCUSSION TITLE
Custom NER training and debug config.cfg file

### DISCUSSION THREAD

**Discussion by akki248** [explosion/spaCy#8268] [Help: Other Questions] [answered]
Hi, 

I'm trying to train a custom NER model.

**Some useful details:**
Python implementation: CPython
Python version       : 3.8.8
IPython version      : 7.22.0
Compiler    : Clang 10.0.0 
OS          : Darwin
Release     : 20.3.0
Machine     : x86_64
Processor   : i386
CPU cores   : 8
Architecture: 64bit
spacy version: 3.0.6

I obtained the config.cfg file after running the command:

`python -m spacy init fill-config base_config.cfg config.cfg`

Then, I make no changes to the generated config.cfg file and run the following command as mentioned here for [debug].(https://spacy.io/api/cli#debug)

`python -m spacy debug model config.cfg ner`

But, I get the following error as shown in the attached snapshot:
<img width="903" alt="Screenshot 2021-06-02 at 3 49 52 PM" src="https://user-images.githubusercontent.com/40318353/120464193-48f0b500-c3ba-11eb-812a-f9725013a869.png">

If I ignore the error and train the custom NER model using the command:
`python -m spacy train config.cfg --output ./output --paths.train ta_train.spacy --paths.dev ta_validation.spacy `

It starts to train (regarding results, I'll start a new discussion). 

I'm unable to understand that if there is an error regarding the tok2vec listener, how am I able to train and how does the output of the tok2vec layer go to the NER transition based parser?

Please let me know if you need any further information from my side. 

Thanks.

**Comment by svlandeg** [comment, marked as answer]
Hi! 

It looks like the generated config file doesn't have valid `paths.train` and `paths.dev` sections, but the `debug model` script needs some example data to initialize the pipeline. 

Could you retry your debug command like this?
```
python -m spacy debug model config.cfg ner --paths.train ta_train.spacy --paths.dev ta_validation.spacy
```

**Reply by akki248** [reply]
Hi,

Thanks. I did retry with the debug command. But I still get the same error as shown in the snapshot. 

![image](https://user-images.githubusercontent.com/40318353/120623310-41deab00-c47d-11eb-970f-bcdd399c7d3d.png)

**Reply by svlandeg** [reply]
Oh, weird. Can you paste me your full config file? Then I can better help debug this.

**Reply by akki248** [reply]
Yes, sure. Here it is:


[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","ner"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 2000
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
ents_per_type = null
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]

**Reply by akki248** [reply]
Hi, I thought of sending a reminder, incase this post has been missed out.

**Reply by svlandeg** [reply]
Hi @akki248 : please don't artificially bump your thread as this is really unhelpful to us. We try to get to everyone's questions in due time, but with no garantuees.

**Reply by svlandeg** [reply]
I tried replicating your error with the config file you pasted. It refers to `@Tokenizers` which should be `@tokenizers`, so I'm guessing that you won't have been able to run this particular config file at all. You may have run another one.

Anyway, when I change L19 to
```
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
```
it runs just fine for me:
```
$ spacy debug model spacy/disc8268.cfg ner
ℹ Using CPU
ℹ To switch to GPU 0, use the option: --gpu-id 0
ℹ Fixing random seed: 0
ℹ Analysing model with ID 60
ℹ Initialized the model with dummy data.
✔ Succesfully ended analysis - model looks good.
```

**Reply by akki248** [reply]
Thanks for the correction. Even after changing the L19 to 
`tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
`

I get the same error when I run the command to debug the ner. It's weird. But, when I run:
`python -m spacy debug model config.cfg tok2vec`

It seems to work. 
![image](https://user-images.githubusercontent.com/40318353/122236460-27c0b600-cedc-11eb-943c-350338e98df7.png)

Some useful details:
Python implementation: CPython
Python version : 3.8.8
IPython version : 7.22.0
Compiler : Clang 10.0.0
OS : Darwin
Release : 20.3.0
Machine : x86_64
Processor : i386
CPU cores : 8
Architecture: 64bit
spacy version: 3.0.6

Do you think there could be something else that I could check?

**Reply by svlandeg** [reply]
I'm afraid I can't be of any further assistance if I don't know which exact config file you've run. I suspect you have multiple config files and you may have been citing the wrong one in this thread, because it's not possible that the one with `@Tokenizers` in it, gave you the error as you said (nor is it possible that the error stayed the same after you changed the typo). The one you cited, worked fine for me (after the typo correction)

Also, as mentioned before, please do not share screenshots of code, but copy and paste the code as text instead (and format it properly with markdown), thanks.

**Reply by akki248** [reply]
OK. Thanks.
